### PR TITLE
MultiProgress: add println and suspend implementation

### PR DESCRIPTION
**Pull request by chris-laplante**
Work-in-progress pull request for implementing println. This does not currently work as nothing is printed for some reason. I have reason to believe it is due to an interaction with e46ca83e, since if you change the line below to `m.println("Starting...\n\n")` it works.

### CODE REVIEW

**Comment by chris-laplante** [comment]
FYI I reverted e46ca83 locally and it fixes the problem. So I guess the goal is to find a way for that commit and this feature to coexist.

**Comment by chris-laplante** [comment]
See #205

**Comment by sigmaSd** [comment]
I tried your branch and it seems to work (I tried the multi example, and it seems to print those lines), is there something needed to reproduce the issue?

I'm on linux if that matters

**Comment by chris-laplante** [comment]
@sigmaSd thanks for giving it a try! I actually did fix the issue, but forgot to mark the pull as ready for review. Sorry about that.

**Comment by sigmaSd** [comment]
@chris-laplante  thanks as well! glad its not a platform issue then, If its just about adding a new line, I guess that's expected with https://github.com/console-rs/indicatif/commit/e46ca83e60dfc5cc64491b26ecd4d361f7e2e1f2, I kind of want it to work that's why I'm on the lookup for issues xD

**Comment by chris-laplante** [comment]
@sigmaSd there's another problem here I didn't initially see. Adding an extra space after the last line (`term.write_str(" ")?;` in `ProgressDrawState::draw_to_term` is not correct in the case of `println`. I noticed this while extending adding `suspend`, which I just pushed. Please give it a try. You will see that the println caused an extra space to be printed out, which is why the "test suspension :)" comes out with a leading space:

```
Starting!
 test suspension :)
```

**Comment by sigmaSd** [comment]
Yes the space hack was definitly the wrong aproach, shifting the missing width increase to the wide elements seems to  be the reasonable choice in #350

File: src/draw_target.rs [hunk 1]
```diff
@@ -309,6 +304,62 @@ impl MultiProgressState {
...
+    pub(crate) fn println<I: AsRef<str>>(&mut self, msg: I) {
+        let mut lines: Vec<String> = msg.as_ref().lines().map(Into::into).collect();
+        let orphan_lines = lines.len();
+
+        if !self.gather_draw_state_lines(&mut lines) {
+            // The draw state consists entirely of the contents of `msg`. Since
+            // `ProgressDrawState::draw_to_term` doesn't add \n to the last line, we need to add it
+            // ourselves. Otherwise the last line of what we print may be overwritten on next draw.
+            if let Some(line) = lines.last_mut() {
+                line.push('\n');
```
Comment (sigmaSd) [inline src/draw_target.rs:348]: is this still needed after https://github.com/console-rs/indicatif/commit/af3bd3154b5f06d3baece6a91eda5dcf48b1e4c0 ?

Comment (sigmaSd) [reply src/draw_target.rs:348]: Removing this line and using https://github.com/console-rs/indicatif/pull/356 seems to make it work

**Comment by djc** [comment]
@chris-laplante did you want to rebase this? I'd like to get the release out soon.

**Comment by chris-laplante** [comment]
> @chris-laplante did you want to rebase this? I'd like to get the release out soon.

Yes, absolutely. I started working on it this morning. Will finish it soon I hope.

**Comment by djc** [comment]
Awesome, thanks!

**Comment by chris-laplante** [comment]
Not ready yet, besides the lint I am still working on diagnosing an issue...

**Comment by chris-laplante** [comment]
@djc not sure if you're getting notifications about the comment I made on my commit (if you are, I apologize for the spam). In case not, I'll repeat it here since I know you want to release soon:

> If you run the 'multi' example that I modified, you'll see that only the first message is printed. The messages for when h1, h2, and h3 finish are not printed. If you change line 247 src/multi.rs to panic instead of return OK, you'll see that it's hitting this line. drawable is returning None here.
> 
> I think the issue is that ProgressDrawTarget::drawable should take the force_draw parameter, rather than Drawable::state. There is a bit of a chicken-and-egg situation going on. You need a DrawStateWrapper to manipulate the force_draw property of ProgressDrawState. But to get one, you need to call ProgressDrawTarget::drawable which (in the case of Term) is beholden to the leaky bucket.

**Comment by djc** [comment]
Sounds reasonable. Please do it as a separate commit (or even separate PR).

File: src/in_memory.rs [hunk 2]
```diff
@@ -54,31 +54,47 @@ impl TermLike for InMemoryTerm {
     }
 
     fn move_cursor_up(&self, n: usize) -> std::io::Result<()> {
-        self.state
-            .lock()
-            .unwrap()
-            .write_str(&*format!("\x1b[{}A", n))
+        match n {
```
Comment (djc) [inline src/in_memory.rs:57]: Did you file a bug against vt100?

File: src/multi.rs [hunk 3]
```diff
@@ -192,6 +215,38 @@ impl MultiState {
...
+    pub(crate) fn println<I: AsRef<str>>(&mut self, msg: I) -> io::Result<()> {
+        let msg = msg.as_ref();
+
+        // If msg is "", make sure a line is still printed
+        let lines: Vec<String> = match msg.is_empty() {
+            false => msg.lines().map(Into::into).collect(),
+            true => vec![String::new()],
+        };
+
+        if self.draw_target.is_hidden() {
```
Comment (djc) [inline src/multi.rs:227]: This isn't needed anymore (`draw_target.drawable()` will return `None` in this case, anyway).

File: src/multi.rs [hunk 4]
```diff
@@ -192,6 +215,38 @@ impl MultiState {
...
+            return Ok(());
+        }
+
+        let mut drawable = match self.draw_target.drawable(true, Instant::now()) {
```
Comment (djc) [inline src/multi.rs:231]: Why can't we call `draw()` here? We should probably deduplicate the code as much as possible.

**Review by djc — COMMENTED** [review]
Mostly looks, some minor nits remaining!

**Review by djc — APPROVED** [review]
LGTM! Good to merge?

**Comment by chris-laplante** [comment]
> LGTM! Good to merge?

Yes please :)

**Comment by chris-laplante** [comment]
Would it be possible to get this merged in today? I'd love to play with this some more over the weekend :)

**Comment by djc** [comment]
Sorry, I somehow missed your previous comment!

**Comment by chris-laplante** [comment]
> Sorry, I somehow missed your previous comment!

No prob, thanks for merging!

**Comment by chris-laplante** [comment]
Any chance of an rc release with these changes?

**Comment by djc** [comment]
rc.8 is out now. 👍

**Comment by chris-laplante** [comment]
> rc.8 is out now. 👍

Thanks! I appreciate it :)

**Pull request merged**